src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs

```diff
@@ -459,65 +461,28 @@ public static partial class JsonSerializer
-        /// <param name="returnType">The type of the object to convert to and return.</param>
-        /// <param name="options">Options to control the behavior during reading.</param>
-        /// <param name="cancellationToken">The <see cref="System.Threading.CancellationToken"/> that can be used to cancel the read operation.</param>
-        /// <exception cref="System.ArgumentNullException">
-        /// <paramref name="utf8Json"/> or <paramref name="returnType"/> is <see langword="null"/>.
-        /// </exception>
-        [RequiresUnreferencedCode(SerializationUnreferencedCodeMessage)]
-        [RequiresDynamicCode(SerializationRequiresDynamicCodeMessage)]
-        public static IAsyncEnumerable<object?> DeserializeAsyncEnumerable(
-            Stream utf8Json,
-            Type returnType,
-            JsonSerializerOptions? options = null,
-            CancellationToken cancellationToken = default)
+        private static IAsyncEnumerable<T> DeserializeAsyncEnumerableCore<T>(Stream utf8Json, JsonTypeInfo<T> jsonTypeInfo, CancellationToken cancellationToken)
         {
-            if (utf8Json is null)
-            {
-                ThrowHelper.ThrowArgumentNullException(nameof(utf8Json));
-            }
-
-            JsonTypeInfo jsonTypeInfo = GetTypeInfo(options, returnType);
-            return jsonTypeInfo.DeserializeAsyncEnumerableAsObject(utf8Json, cancellationToken);
-        }
+            Debug.Assert(jsonTypeInfo.IsConfigured);
+            jsonTypeInfo.AsyncEnumerableQueueTypeInfo ??= CreateQueueTypeInfo(jsonTypeInfo);
+            return jsonTypeInfo.DeserializeAsyncEnumerable(utf8Json, cancellationToken);
 
-        /// <summary>
-        /// Wraps the UTF-8 encoded text into an <see cref="IAsyncEnumerable{Object}" />
-        /// that can be used to deserialize root-level JSON arrays in a streaming manner.
-        /// </summary>
-        /// <returns>An <see cref="IAsyncEnumerable{Object}" /> representation of the provided JSON array.</returns>
-        /// <param name="utf8Json">JSON data to parse.</param>
-        /// <param name="jsonTypeInfo">Metadata about the element type to convert.</param>
-        /// <param name="cancellationToken">The <see cref="System.Threading.CancellationToken"/> that can be used to cancel the read operation.</param>
-        /// <exception cref="System.ArgumentNullException">
-        /// <paramref name="utf8Json"/> or <paramref name="jsonTypeInfo"/> is <see langword="null"/>.
-        /// </exception>
-        public static IAsyncEnumerable<object?> DeserializeAsyncEnumerable(
-            Stream utf8Json,
-            JsonTypeInfo jsonTypeInfo,
-            CancellationToken cancellationToken = default)
-        {
-            if (utf8Json is null)
-            {
-                ThrowHelper.ThrowArgumentNullException(nameof(utf8Json));
-            }
-            if (jsonTypeInfo is null)
+            static JsonTypeInfo<Queue<T>> CreateQueueTypeInfo(JsonTypeInfo<T> jsonTypeInfo)
```

Comment:
@jkotas @MichalStrehovsky after a lot of experimentation, I was able to pin down the AOT recursion warnings to this method being defined within `JsonTypeInfo<T>` itself, regardless of whether it's static or not. I find that somewhat surprising, any chance we could be looking at a false positive in the AOT compiler?

Reply:
It is limitation of AOT compiler analysis. Here is minimal repro:
```
Console.WriteLine(new G<string>().M());

class G<T>
{
    public object M()
    {
        return new G<List<T>>();
    }
}
```

When the `G<List<T>>` instance is created in `M()`, we assume that it is possible for `M()` to be called on that instance again, and that triggers the infinite generic recursion. I have opened https://github.com/dotnet/runtime/issues/85184 on this.